docs(proposal): monitoring of custom resources (#133)
strimzi/proposals: 086-monitoring-of-custom-resources.md

# Monitoring of custom resources

In https://github.com/strimzi/strimzi-kafka-operator/issues/9802 there was a discussion about how to monitor the state of a custom resources (CR) like `KafkaTopic`, `KafkaUser`, ... .
For a simple and Kubernetes native monitoring, [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) (KSM) can be used which utilizes e.g. the status object of a Kubernetes resource.

## Current situation

All Strimzi operators (Cluster, User and Topic operators) provide metrics in Prometheus format.
However, currently, only the Cluster Operator provides a metric (`strimzi_resource_state`) describing the state of the custom resources (whether they are ready or not).
It provides this metric for the `Kafka`, `KafkaConnect`, `KafkaBridge`, `KafkaMirrorMaker`, `KafkaMirrorMaker2`, and `KafkaRebalance` custom resources.

> Currently, it does not provide this metric for `KafkaNodePool`, `KafkaConnector`, and `StrimziPodSet` custom resources.

Reviewer: I am not sure we want to mention (and have) the `StrimziPodSet` custom resource to be "monitored" or exposed through metrics, it's an internal resource.

Reviewer: User should always interact via the API which is represented by the `Kafka` prefixed resources. They never touch a `StrimziPodSet`.

Reviewer 2: The StrimziPodSets are part of our metrics. So I do not see why should they be excluded from here.

Other tools like Flux handles the monitoring of CRs via KSM like documented in their [documentation](https://fluxcd.io/flux/monitoring/custom-metrics/).
In addition, Flux provides an [example repository](https://github.com/fluxcd/flux2-monitoring-example) for configuring KSM as part of [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.
But the required configuration is just being part of the ConfigMap used by KSM as shown by @im-konge in this [issue](https://github.com/strimzi/strimzi-kafka-operator/issues/10276#issuecomment-2276088493).

## Proposal

So in order to have a monitoring in place for all CRs and implementing them in a scalable and Kubernetes native way KSM can be used.
KSM is (mostly) deployed via a simple Kubernetes `Deployment` and querries the Kubernetes api for e.g. the status object of a Kubernetes resource indicating that the CR having a problem or contains a deprecated configuration.
This can also be done for CRs exclusively by using the `--custom-resource-state-config` flag which is helpful for e.g. `KafkaTopic`, `KafkaUser`, ... CRs.

> During the initial phase when implementing this proposal, the support in the examples will be implemented at least for the existing resources for which Strimzi publishes already existing metric(s), additionally for `KafkaUser` and `KafkaTopic` custom resources.

Reviewer: What does it exactly means? Several PRs to cover multiple CRs or that you'll need another proposal (why?) for `KafkaUser` and `KafkaTopic`?

Author: That was a suggestion from @scholzj , so I would that for him :D

Reviewer: I am fine, I am just asking what does it exactly mean ... if several proposals or several PRs to address stuff step by step.

Reviewer: Thanks Jakub. @sebastiangaiser I would make the sentence clearer by using part of the above explanation.

The ConfigMap should look like e.g.:

```
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
  name: strimzi-kube-state-metrics-config
data:
  config.yaml: |
    spec:
      resources:
      - groupVersionKind:
          group: kafka.strimzi.io
          version: v1beta2
          kind: KafkaTopic
        metricNamePrefix: strimzi_kafka_topic
        metrics:
          - name: resource_info
            help: "The current state of a Strimzi kafka topic resource"
            each:
              type: Info
              info:
                labelsFromPath:
                  name: [ metadata, name ]
            labelsFromPath:
              exported_namespace: [ metadata, namespace ]
              partitions: [ spec, partitions ]
              replicas: [ spec, replicas ]
              ready: [ status, conditions, "[type=Ready]", status ]
              generation: [ status, observedGeneration ]
              topicId: [ status, topicId ]
              topicName: [ status, topicName ]
      - ...
```

> For YAML users the ConfigMap works fine when creating a dedicated deployment.

Reviewer 3: I don't quite grok what this sentence means. What is a YAML user? Deployment of what?

This works also fine when implement the configuration inside [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.

For (all other) Helm users, this should be left to the user implementing the KSM based monitoring, as there are many different Helm charts but all of them will handle the same ConfigMap in the end.
This could be explained in a Blog Post with a static version of the [prometheus-community kube-state-metrics Helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics/) at least once.

> In addition, example `PrometheusRules` should be defined for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.

Reviewer 3: Can we make it clearer where those rules should be defined? I think you mean this:

```suggestion
In addition, Strimzi will define example `PrometheusRules` for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.
```

[This PR](https://github.com/strimzi/strimzi-kafka-operator/pull/10277) could be used as an implementation idea.

## Affected/not affected projects

The only affected projects are:
- CO

## Compatibility

Strimzi provides the `strimzi_resource_state` metric(s) implemented in CO.

> The previously elaborated metrics(s) should be deprecated and replaced in favour of KSM based metrics.

Author: ```suggestion
The previously elaborated metrics(s) should be deprecated and replaced in favor of KSM based metrics which means that KSM will be responsible for providing these metrics instead of CO.
Due to this change, the metric names, labels and format will change.
```

Reviewer 2: Maybe you can provide an example of how the new metrics would look like (similarly to the example you already have here with how the current metrics look like)? That might give more clarity on what is changing and how. (but in general, the fact that the metrics won't be the same is why the original metrics will be deprecated first and removed only later, to give people time to adjust to it, so at least for me this is expected and not a problem)

Author: Will add them

This also applies for PrometheusRules which should be replaced.
The proposed way would be implementing the KSM and deprecating the current metrics in Strimzi CO in version 0.45 and removing them in version 0.49 to give users enough time adjusting their monitoring if needed.
So there is no immediate impact for users and enough time for migration.

> 

Reviewer 3: I think you should say explicitly that this proposal is changing what's responsible for providing the metrics, but that the metrics provided by KSM will be compatible with the metrics that were provided by the CO. I.e. same metrics names and tags, measuring the same things.

I guess one question I have is for people who are using Strimzi and the metrics it currently provides, but who might not themselves have the ability to enable the addon required by KSM. Does anyone know how commonly the addon is enabled in pratice?

Author: > I guess one question I have is for people who are using Strimzi and the metrics it currently provides, but who might not themselves have the ability to enable the addon required by KSM. Does anyone know how commonly the addon is enabled in pratice?

I'm not sure if I got you here but KSM is neither an addon to Kubernetes nor to Strimzi. It's an application exposing the status (and other parts) of Kubernetes resources. So you could deploy KSM as a standalone application with access to the Kubernetes API server 😉

Author: Ah [they specify](https://github.com/kubernetes/kube-state-metrics) it as `Add-on agent to generate and expose cluster-level metrics.`, sorry for that. But in the end you could e.g. have that for multiple different tools using custom resources. E.g. one KSM deployment for Strimzi, one for Flux, ...

For (all other) Helm users, this should be left to the user implementing the KSM based monitoring, as there are many different Helm charts but all of them will handle the same ConfigMap in the end.
This could be explained in a Blog Post with a static version of the [prometheus-community kube-state-metrics Helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics/) at least once.

In addition, Strimzi will define example `PrometheusRules` for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.
[This PR](https://github.com/strimzi/strimzi-kafka-operator/pull/10277) could be used as an implementation idea.

## Affected/not affected projects

The only affected projects are:
- CO

## Compatibility

Strimzi provides the `strimzi_resource_state` metric(s) implemented in CO.

The previously elaborated metrics(s) should be deprecated and replaced in favor of KSM based metrics which means that KSM will be responsible for providing these metrics instead of CO.

> Due to this change, the metric names, labels and format will change.

Author: ```suggestion
Due to this change, the metric names, labels and format will change.
Example for a `KafkaUser` metric:

```
strimzi_kafka_user_resource_info{customresource_group="kafka.strimzi.io",customresource_kind="KafkaUser",customresource_version="v1beta2",exported_namespace="kafka",generation="6",name="my-kafka-user",ready="True",secret="my-kafka-user",username="CN=my-kafka-user"} 1`
```

Example for a `KafkaTopic` metric:

```
strimzi_kafka_topic_resource_info{customresource_group="kafka.strimzi.io",customresource_kind="KafkaTopic",customresource_version="v1beta2",exported_namespace="kafka",generation="1",name="my-kafka-topic",partitions="12",ready="True",replicas="3",topicId="asdfasdfasdfasdf",topicName="my-kafka-topic"} 1
```

```

For (all other) Helm users, this should be left to the user implementing the KSM based monitoring, as there are many different Helm charts but all of them will handle the same ConfigMap in the end.
This could be explained in a Blog Post with a static version of the [prometheus-community kube-state-metrics Helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics/) at least once.

In addition, Strimzi will define example `PrometheusRules` for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.
[This PR](https://github.com/strimzi/strimzi-kafka-operator/pull/10277) could be used as an implementation idea.

## Affected/not affected projects

The only affected projects are:
- CO

## Compatibility

Strimzi provides the `strimzi_resource_state` metric(s) implemented in CO.

The previously elaborated metrics(s) should be deprecated and replaced in favor of KSM based metrics which means that KSM will be responsible for providing these metrics instead of CO.

> Due to this change, the metric names, labels and format will change.

Reviewer: I see you added two examples of newer metrics for `KafkaUser` and `KafkaTopic`. Because here you say "Due to this change, the metric names, labels and format will change." so referring to the existing ones, could it be useful to have an example of how already existing metrics change?

Author: @ppatierno I can try to add one later today for a `Kafka` resources but it probably won't have all the fields which are possible.
As a rule of thumb the pattern would be `strimzi_` + `custom_resource` + `_resource_info{all_exposed_fields}`. I will add the pattern, too.

Author: Shall I still add one for `Kafka` custom resources?

Reviewer: I would say yes, just one concrete example, mostly because the pattern has this `all_exposed_fields` but it would be great to see even some of them in one example.

Reviewer: +1 I am totally fine with a `Kafka` related example.

Author: Done

> Due to this change, the metric names, labels and format will change.

Author: ```suggestion
Due to this change, the metric names, labels and format will change.

The pattern for new metrics will look like `strimzi_` + `custom_resource` + `_resource_info{all_exposed_fields}`

```

This could be explained in a Blog Post with a static version of the [prometheus-community kube-state-metrics Helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics/) at least once.

In addition, Strimzi will define example `PrometheusRules` for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.
[This PR](https://github.com/strimzi/strimzi-kafka-operator/pull/10277) could be used as an implementation idea.

## Affected/not affected projects

The only affected projects are:
- CO

## Compatibility

Strimzi provides the `strimzi_resource_state` metric(s) implemented in CO.

The previously elaborated metrics(s) should be deprecated and replaced in favor of KSM based metrics which means that KSM will be responsible for providing these metrics instead of CO.
Due to this change, the metric names, labels and format will change.

> The pattern for new metrics will look like `strimzi_` + `custom_resource` + `_resource_info{all_exposed_fields}`

Reviewer: ```suggestion
The pattern for new metrics will look like `strimzi_` + `<custom_resource>` + `_resource_info{<all_exposed_fields>}`
```

However, currently, only the Cluster Operator provides a metric (`strimzi_resource_state`) describing the state of the custom resources (whether they are ready or not).
It provides this metric for the `Kafka`, `KafkaConnect`, `KafkaBridge`, `KafkaMirrorMaker`, `KafkaMirrorMaker2`, and `KafkaRebalance` custom resources.
Currently, it does not provide this metric for `KafkaNodePool`, `KafkaConnector`, and `StrimziPodSet` custom resources.
The User and Topic Operators currently do not provide this metric for the `KafkaUser` and `KafkaTopic` custom resources.
Examples:

```
$ curl -s localhost:8080/metrics | grep strimzi_resource_state
# HELP strimzi_resource_state Current state of the resource: 1 ready, 0 fail
# TYPE strimzi_resource_state gauge
strimzi_resource_state{kind="Kafka",name="my-cluster",reason="none",resource_namespace="myproject",} 1.0
strimzi_resource_state{kind="KafkaConnect",name="my-connect",reason="none",resource_namespace="myproject",} 1.0
```

## Motivation

Strimzi already provides example configurations being compatible with the prometheus-operator via e.g. `PodMonitor` for scraping the metrics and e.g. `PrometheusRules` for alerting.

> But in general, if a deployed Kubernetes resource is having a problem or a deprecated configuration is used, you want to get notified about that and can fix the problem, afterward.

Reviewer 4: ```suggestion
In general, you want to be notified if a deployed Kubernetes resource encounters a problem or uses a deprecated configuration, so you can address the issue promptly.
```

> Other tools like Flux handles the monitoring of CRs via KSM like documented in their [documentation](https://fluxcd.io/flux/monitoring/custom-metrics/).

Reviewer 4: ```suggestion
Other tools, such as Flux, handle the monitoring of custom resources (CRs) via Kubernetes Service Monitor (KSM), as described in their [documentation](https://fluxcd.io/flux/monitoring/custom-metrics/).
```

Author: ```suggestion
Other tools, such as Flux, handle the monitoring of custom resources (CRs) via KSM, as described in their [documentation](https://fluxcd.io/flux/monitoring/custom-metrics/).
```
@PaulRMellor thank you for the suggestion.
KSM is the abbreviation for kube-state-metrics as declared in line 4. What do you think about this?

In addition, Flux provides an [example repository](https://github.com/fluxcd/flux2-monitoring-example) for configuring KSM as part of [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.
But the required configuration is just being part of the ConfigMap used by KSM as shown by @im-konge in this [issue](https://github.com/strimzi/strimzi-kafka-operator/issues/10276#issuecomment-2276088493).

## Proposal

So in order to have a monitoring in place for all CRs and implementing them in a scalable and Kubernetes native way KSM can be used.

> KSM is (mostly) deployed via a simple Kubernetes `Deployment` and querries the Kubernetes api for e.g. the status object of a Kubernetes resource indicating that the CR having a problem or contains a deprecated configuration.

Reviewer 4: ```suggestion
KSM is (mostly) deployed via a simple Kubernetes `Deployment` and queries the Kubernetes api for e.g. the status object of a Kubernetes resource indicating that the CR having a problem or contains a deprecated configuration.
```

This can also be done for CRs exclusively by using the `--custom-resource-state-config` flag which is helpful for e.g. `KafkaTopic`, `KafkaUser`, ... CRs.

> During the initial phase when implementing this proposal, the support in the examples will be implemented at least for the existing resources for which Strimzi publishes already existing metric(s), additionally for `KafkaUser` and `KafkaTopic` custom resources.

Reviewer 4: ```suggestion
During the initial phase when implementing this proposal, the support in the examples will be implemented at least for the existing resources for which Strimzi publishes already existing metrics, additionally for `KafkaUser` and `KafkaTopic` custom resources.
```

The ConfigMap should look like e.g.:

```
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
  name: strimzi-kube-state-metrics-config
data:
  config.yaml: |
    spec:
      resources:
      - groupVersionKind:
          group: kafka.strimzi.io
          version: v1beta2
          kind: KafkaTopic
        metricNamePrefix: strimzi_kafka_topic
        metrics:
          - name: resource_info
            help: "The current state of a Strimzi kafka topic resource"
            each:
              type: Info
              info:
                labelsFromPath:
                  name: [ metadata, name ]
            labelsFromPath:
              exported_namespace: [ metadata, namespace ]
              partitions: [ spec, partitions ]
              replicas: [ spec, replicas ]
              ready: [ status, conditions, "[type=Ready]", status ]
              generation: [ status, observedGeneration ]
              topicId: [ status, topicId ]
              topicName: [ status, topicName ]
      - ...
```

For users deploying Kubernetes manifests as plain YAML, the ConfigMap works fine when creating a dedicated deployment of KSM.

> This works also fine when implement the configuration inside [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.

Reviewer 4: ```suggestion
This works also fine when implementing the configuration inside [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.
```

For (all other) Helm users, this should be left to the user implementing the KSM based monitoring, as there are many different Helm charts but all of them will handle the same ConfigMap in the end.
This could be explained in a Blog Post with a static version of the [prometheus-community kube-state-metrics Helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics/) at least once.

> In addition, Strimzi will define example `PrometheusRules` for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.

Reviewer 4: ```suggestion
In addition, Strimzi will define example `PrometheusRules` for all CRs having issues or using deprecated configuration in a simple way, so that the Cluster Operator look for an error message in the status object of the CR.
```

# Monitoring of custom resources

In https://github.com/strimzi/strimzi-kafka-operator/issues/9802 there was a discussion about how to monitor the state of a custom resources (CR) like `KafkaTopic`, `KafkaUser`, ... .
For a simple and Kubernetes native monitoring, [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) (KSM) can be used which utilizes e.g. the status object of a Kubernetes resource.

## Current situation

> Strimzi provides metrics for several use cases but currently not for the state of `KafkaTopic` and `KafkaUser` CRs.

Reviewer 2: Maybe we can expand this a bit?

```suggestion
All Strimzi operators (Cluster, User and Topic operators) provide metrics in Prometheus format.
However, currently, only the Cluster Operator provides a metric (`strimzi_resource_state`) describing the state of the custom resources (whether they are ready or not).
It provides this metric for the `Kafka`, `KafkaConnect`, `KafkaBridge`, KafkaMirrorMaker`, `KafkaMirrorMaker2`, and `KafkaRebalance` resources.
It does not currently provide this metric for `KafkaNodePool`, `KafkaConnector`, and `StrimziPodSet` custom resources.
The User and Topic Operators currently do not provide this metric for the `KafkaUser` and `KafkaTopic` resources.
```

> TODO need input for the status of other CRs and their metrics, e.g. `Kafka`, `StrimziPodSet`, `KafkaRebalance`...

Reviewer 2: The metric name is `strimzi_resource_state`. It is currently issued only by the Cluster Operator and only for the user level resources (e.g. Kafka, KafkaConnect ... but not for StrimziPodSet resources). If you wanna include some examples here:

```
$ curl -s localhost:8080/metrics | grep strimzi_resource_state
# HELP strimzi_resource_state Current state of the resource: 1 ready, 0 fail
# TYPE strimzi_resource_state gauge
strimzi_resource_state{kind="Kafka",name="my-cluster",reason="none",resource_namespace="myproject",} 1.0
strimzi_resource_state{kind="KafkaConnect",name="my-connect",reason="none",resource_namespace="myproject",} 1.0
```

Or as a screenshot:
![Screenshot 2024-09-23 at 16 19 56](https://github.com/user-attachments/assets/d08ad287-f1cf-4012-a5d3-ca612a4d6c18)

## Motivation

Strimzi already provides example configurations being compatible with the prometheus-operator via e.g. `PodMonitor` for scraping the metrics and e.g. `PrometheusRules` for alerting.

> This is implemented by metrics-exporter via JMX or kafka-exporter.

Reviewer 2: Actually, as we are talking about operators here, they handle it on their own. So I would probably delete this sentence.

But in general, if a deployed Kubernetes resource is having a problem or a deprecated configuration is used, you want to get notified about that and can fix the problem, afterwards.
Other tools like Flux handles the monitoring of CRs via KSM like documented in their [documentation](https://fluxcd.io/flux/monitoring/custom-metrics/).
In addition Flux provides an [example repository](https://github.com/fluxcd/flux2-monitoring-example) for configuring KSM as part of [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.
But the required configuration is just being part of the ConfigMap used by KSM as shown by @im-konge in this [issue](https://github.com/strimzi/strimzi-kafka-operator/issues/10276#issuecomment-2276088493).

## Proposal

So in order to have a monitoring in place for all CRs and implementing them in a scalable and Kubernetes native way KSM can be used.
KSM is (mostly) deployed via a simple Kubernetes `Deployment` and querries the Kubernetes api for e.g. the status object of a Kubernetes resource indicating that the CR having a problem or contains a deprecated configuration.

> This can also be done for CRs exclusivly by using the `--custom-resource-state-config` flag which is helpful for e.g. `KafkaTopic`, `KafkaUser`, ... CRs.

Reviewer 2: Maybe you should clarify that in the initial phase, the support in the examples will be implemented at least for the existing resources for which we publish the existing metric + for `KafkaUser` and `KafkaTopic`.

### Deployment

Strimzi should provide guidance for existing fields in status object in order to indicate if a CR is having a problem or uses a deprecated configuration in the official documentation.
By this users can extend monitoring of CRs on their own requirements.

> In general, an example ConfigMap should be provided in the examples directory of [strimzi/strimzi-kafka-operator](https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples) specifying a good starting configuration.

Reviewer 2: Should we be more specific here? I guess it will be somewhere as `https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples/metrics/kube-state-metrics` probably?

Maybe you could include a simple sample of what the config map will look like to give a better idea? E.g. just for a single resource.

For YAML users the ConfigMap works fine when creating a dedicated deployment.

This works also fine when implement the configuration inside of [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) Helm chart.

For (all other) Helm users, this should be left to the user implementing the KSM based monitoring, as there are many different Helm charts but all of them will handle the same ConfigMap in the end.
This could be explained in a Blog Post with a static version of the [prometheus-community kube-state-metrics Helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics/) at least once.

In addition, example `PrometheusRules` should be defined for all CRs for having a problem or uses a deprecated configuration in a simple way, so that a cluster operator can have a look for an error message in the status object of the CR.
[This PR](https://github.com/strimzi/strimzi-kafka-operator/pull/10277) could be used as an implementation idea.

> ## Affected/not affected projects

Reviewer 2: Did you checked if the current metric is used in any Grafana Dashboards? I think it is not, so there should be nothing to worry about, but I did not searched for it really, just basing it on my memory.

Author: When searching the [packaging/helm-chart path](https://github.com/search?q=repo%3Astrimzi%2Fstrimzi-kafka-operator%20path%3A%2F%5Epackaging%5C%2Fhelm-charts%5C%2Fhelm3%5C%2F%2F%20strimzi_resource_state&type=code) I didn't found anything related to `strimzi_resource_state` metric.

The only affected projects are:
- CO
- TO

> - UO

Reviewer 2: I think the Cluster Operator is the only affected project -> the others do not provide such a metric so nothing will change there.

## Compatibility

> TODO need input about already implemented metrics, e.g. for `Kafka` resources

Reviewer 2: I think what you have in the paragraph below is mostly good. In general:
* You should make it clear that there is no immediate impact right away and users will have the 4 releases to adjust
* That after the 4 releases, the `strimzi_resource_state` metric will be removed from the Cluster Operator completely

The previously elaborated metrics should be deprecated and replaced in favour of KSM based metrics.
This also applies for PrometheusRules which should be replaced.

> The proposed way would be implementing the KSM and deprecating the current metrics in Strimzi CO/TO/UO in version XX and removing them in version XZ to give users enough time (2 releases) adjusting their monitoring if needed.

Reviewer 2: I guess I would leave 4 releases? 2 are quite a short time period. Also, it should be only the Cluster Operator that gets mentioned here. I would also fill in some expected versions here. Maybe 0.45 for the deprecation and 0.49 for the removal. if the proposal gets approved and implemented earlier, we can simply update it.
